index-scheduler/src/batch.rs

```diff
@@ -416,41 +412,46 @@ impl IndexScheduler {
             )));
         }
 
-        // 5. We take the next task and try to batch all the tasks associated with this index.
-        if let Some(task_id) = enqueued.min() {
-            let task = self.get_task(rtxn, task_id)?.ok_or(Error::CorruptedTaskQueue)?;
+        // 5. We make a batch from the unprioritised tasks. Start by taking the next enqueued task.
+        let task_id = if let Some(task_id) = enqueued.min() { task_id } else { return Ok(None) };
+        let task = self.get_task(rtxn, task_id)?.ok_or(Error::CorruptedTaskQueue)?;
 
-            // This is safe because all the remaining task are associated with
-            // AT LEAST one index. We can use the right or left one it doesn't
-            // matter.
-            let index_name = task.indexes().unwrap()[0];
-            let index_already_exists = self.index_mapper.exists(rtxn, index_name)?;
-
-            let index_tasks = self.index_tasks(rtxn, index_name)? & enqueued;
-
-            // If autobatching is disabled we only take one task at a time.
-            let tasks_limit = if self.autobatching_enabled { usize::MAX } else { 1 };
-
-            let enqueued = index_tasks
-                .into_iter()
-                .take(tasks_limit)
-                .map(|task_id| {
-                    self.get_task(rtxn, task_id)
-                        .and_then(|task| task.ok_or(Error::CorruptedTaskQueue))
-                        .map(|task| (task.uid, task.kind))
-                })
-                .collect::<Result<Vec<_>>>()?;
-
-            if let Some((batchkind, create_index)) =
-                crate::autobatcher::autobatch(enqueued, index_already_exists)
-            {
-                return self.create_next_batch_index(
-                    rtxn,
-                    index_name.to_string(),
-                    batchkind,
-                    create_index,
-                );
-            }
+        // 5.1. If the task is an index swap, create the batch directly.
+        if matches!(task.kind.as_kind(), Kind::IndexSwap) {
+            return Ok(Some(Batch::IndexSwap { task }));
+        }
+
+        // 5.2. Otherwise, get the index name associated with the task and use the autobatcher
+        // to batch the enqueued tasks associated with this index.
+
+        // This is safe because all the remaining task are associated with an index.
+        let index_name = task.index_uid().unwrap();
```

Comment:
Wouldn't it be better to use this unwrap to our advantage and match on the output of `index_uid`, here?
Like checking that it returns an index uid and if not check that it is, indeed, an IndexSwap.

Reply:
The initial goal was to always create the IndexSwap batch manually here, even when it is associated with some indexes. But now I realised that we can't remove `IndexSwap` from the `BatchKind` type, since we need to stop batching tasks together when the autobatcher sees an IndexSwap task.

So now I have reverted the autobatcher to what it was, and I will only create the `IndexSwap` batch manually in the `create_batch` method when it is empty.